File: snippets/ts/src/foundations/services/workflow.ts
import * as restate from "@restatedev/restate-sdk";

function createUserEntry(req: { userId: string; user: { name: string; email: string }}) {
  return undefined;
}

function sendVerificationEmail(param: {
  user: { name: string; email: string };
  secret: string;
}) {
  return undefined;
}

// <start_here>
const signupWorkflow = restate.workflow({
  name: "UserSignup",
  handlers: {
    run: async (ctx: restate.WorkflowContext, user: { name: string; email: string }) => {
      // workflow ID = user ID; workflow runs once per user
      const userId = ctx.key;

      await ctx.run("create", () => createUserEntry({ userId, user }));

      const secret = ctx.rand.uuidv4();
      await ctx.run("mail", () => sendVerificationEmail({ user, secret }));

      const clickSecret = await ctx.promise<string>("email-link-clicked");
      return clickSecret === secret;
    },

    click: async (ctx: restate.WorkflowSharedContext, request: { secret: string }) => {
      await ctx.promise<string>("email-link-clicked").resolve(request.secret);
    },
  },
});
// <end_here>

export default signupWorkflow;

File: snippets/python/src/foundations/services/workflow.py
import restate
from pydantic import BaseModel
from restate import WorkflowContext, WorkflowSharedContext
from typing import Dict
import uuid


class User(BaseModel):
    name: str
    email: str


def create_user_entry(user_id: str, user: User) -> None:
    return None


def send_verification_email(user: User, secret: str) -> None:
    return None


# <start_here>
signup_workflow = restate.Workflow("UserSignup")


@signup_workflow.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    # workflow ID = user ID; workflow runs once per user
    user_id = ctx.key()

    await ctx.run_typed("create", create_user_entry, user_id=user_id, user=user)

    secret = str(ctx.uuid())
    await ctx.run_typed("mail", send_verification_email, user=user, secret=secret)

    click_secret = await ctx.promise("email-link-clicked", type_hint=str).value()
    return click_secret == secret


@signup_workflow.handler()
async def click(ctx: WorkflowSharedContext, secret: str) -> None:
    await ctx.promise("email-link-clicked", type_hint=str).resolve(secret)


# <end_here>
